File: 二叉树/437.路径总和III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        if(root==null) return 0;
        Queue<TreeNode> que = new LinkedList<>();
        que.add(root);
        while(!que.isEmpty()){
            int size= que.size();
            for(int i=0;i<size;i++){
                TreeNode tmp = que.poll();
                dfs(tmp,targetSum,0);
                if(tmp.left!=null){
                    que.add(tmp.left);
                }
                if(tmp.right!=null){
                    que.add(tmp.right);
                }
            }
        }
        return cnt;
    }

    int cnt=0;
    public void dfs(TreeNode root,int targetSum,long sum){
        if(root==null){
            return;
        }
        sum+=root.val;
        if(sum==targetSum){
            cnt++;
        }
        dfs(root.left,targetSum,sum);
        dfs(root.right,targetSum,sum);
        sum-=root.val;
    }
}

File: 二叉树/437.路径总和III.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	sum, cnt := 0, 0
	var dfs func(*TreeNode, int, int)
	dfs = func(root *TreeNode, targetSum int, sum int) {
		if root == nil {
			return
		}
		sum += root.Val
		if sum == targetSum {
			cnt++
		}
		dfs(root.Left, targetSum, sum)
		dfs(root.Right, targetSum, sum)
	}

	que := []*TreeNode{root}
	for len(que) > 0 {
		size := len(que)
		for i := 0; i < size; i++ {
			tmp := que[i]
			dfs(tmp, targetSum, sum)
			if tmp.Left != nil {
				que = append(que, tmp.Left)
			}
			if tmp.Right != nil {
				que = append(que, tmp.Right)
			}
		}
		que = que[size:]
	}
	return cnt
}   

